Allow bi-directional Insteon device links

geoffdavis opened the pull request:
The underlying insteonplm module supports adding devices with
bidirectional communication between the PLM and the device (linking mode
3). This change exposes that functionality.

## Breaking Change:

<!-- What is breaking and why we have to break it. Remove this section only if it was NOT a breaking change. -->

## Description:


**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

geoffdavis committed:
Allow bi-directional Insteon device links

The underlying insteonplm module supports adding devices with
bidirectional communication between the PLM and the device (linking mode
3). This change exposes that functionality.

geoffdavis committed:
quoted description string for add_all_link

MartinHjelmare commented on homeassistant/components/insteon/__init__.py:
```diff
@@ -324,7 +327,15 @@ def add_all_link(service):
         group = service.data.get(SRV_ALL_LINK_GROUP)
         mode = service.data.get(SRV_ALL_LINK_MODE)
-        link_mode = 1 if mode.lower() == SRV_CONTROLLER else 0
+        if mode.lower() == SRV_CONTROLLER:
```
Please move the lowercase modification to the service schema. Use `vol.Lower`.

geoffdavis replied:
I'm not quite sure what you mean here. The schema entries themselves are already lowercase (lines 89-91 in the PR). Are you saying to take the results of service.data.get and lowercase that with voluptuous.utils.Lower()?

Since the schema has already validated the contents of mode, can I just remove the calls to str.lower() entirely, or is voluptuous not case-sensitive? Forgive me, it's not a module I've worked with before.

MartinHjelmare replied:
Yeah, if the mode string is already lowercase we don't need to make it lowercase after the `vol.In` check. We could add `vol.Lower` before this check to allow users to be less specific with the casing, but it's not strictly necessary.

MartinHjelmare commented on homeassistant/components/insteon/services.yaml:
```diff
@@ -2,16 +2,16 @@ add_all_link:
       example: 1
     mode:
-      description: Linking mode   controller - IM is controller   responder - IM is responder
+      description: "Linking mode:  controller - IM is controller   responder - IM is responder    both - IM is both controller and responder"
```
This reads very weird? I suggest using comma to separate the different example options, not multiple spaces.

geoffdavis replied:
I agree, it's weird layout. I wasn't sure if it was my place to fix it, since it wasn't my code. I'll see if I can clean it up a bit.

geoffdavis committed:
Remove str.lower() call

Schema validation forces the input to be the correct case, so no need to
coerce the string any further.

geoffdavis committed:
Clean up verbiage for add_all_link mode parameter

geoffdavis committed:
Make example match insteonplm module default

The underlying python_insteonplm module users 0 as the default for the
group number. Change the example value to match the default.

MartinHjelmare commented:
Just ping me and I'll merge if we're happy as it is now.

geoffdavis commented:
I’m happy with it - let’s go ahead and merge. Thanks @MartinHjelmare

MartinHjelmare commented:
This service is described in the docs. Please update the docs and link that PR in the description of this PR. Then we can merge.

https://www.home-assistant.io/integrations/insteon/#linking-devices-to-the-insteon-modem

springstan commented:
@geoffdavis any updates on this? Please update the docs accordingly and resolve the merge conflict, thanks 👍

geoffdavis commented:
After working with the code as it ships currently for a while, I have yet to find a situation where I would want to create bi-directional links with the same scene id on both ends, which is what the "both" mode does. Typically, the link database on the PLM uses some high value scene ID for housekeeping links, and reserves the low link IDs for user-created Insteon scenes. Meanwhile, the light switches and other devices use only a few very low scene IDs

As such, I don't think this is worth merging now. Instead, we ought to focus on automatically creating the correct links for a new device when it's added to the PLM - such as the battery low link, the light switch pressed link, etc. These vary on a per-device basis, but would be very handy to have in the home assistant code base.
